### DISCUSSION TITLE
Re-initializing state between token routes?

### DISCUSSION THREAD

**Discussion by rfedoruk** [TanStack/router#1388] [Q&A] [answered]
So I'm trying to see if this is the best way to do this. 

I am using file based routing, and I have a route that is say: $postId.tsx
In that route, I have some react state that I am initializing with useState(). In that Post I am modifying this state. When I switch to a post with a different postId that state does not re-initialiize. 

The way around this that I found is using a useEffect with the dependency array of Route Param.
I don't know if that is a bug or not but my assumption was that it would? Is there a cleaner or better way to do this?

Any help would be appreciated, thank you!
here is the link to stackblitz (posts.$postId.tsx file has what I am talking about,):
https://stackblitz.com/edit/tanstack-router-786xuh?file=src%2Froutes%2Fposts.%24postId.tsx

also here is a short snippet: 
```function PostComponent() {
  const post = Route.useLoaderData();
  const params = Route.useParams();

  const [counter, setCounter] = React.useState(0);

  // Uncomment this to have counter reset between routes.
  // React.useEffect(() => {
  //   setCounter(0);
  // }, [params.postId]);

  return (
    <div className="space-y-2">
      <h4 className="text-xl font-bold underline">{post.title}</h4>
      {counter}
      <button onClick={() => setCounter((prev) => prev + 1)}>
        Increase Count
      </button>
</div> )
}
```

**Comment by SeanCassiere** [comment, marked as answer]
I'd probably stay away from the `useEffect` and use `key` to trigger the re-render.

```tsx
export const Route = createFileRoute('/posts/$postId')({
  loader: async ({ params: { postId } }) => fetchPost(postId),
  errorComponent: PostErrorComponent as any,
  notFoundComponent: () => {
    return <p>Post not found</p>;
  },
  component: () => {
    const { postId } = Route.useParams();
    return <PostComponent key={postId} />;
  },
});
```

**Reply by rfedoruk** [reply]
This works great, thank you!

**Reply by VladKopylov** [reply]
@SeanCassiere is there any way to get the parameter without using a hook?
I like this solution for resetting state of routes. But question how to manage this eslint rule (do not suggest disabling eslint :) )
![Знімок екрана 2024-10-24 о 13 55 46](https://github.com/user-attachments/assets/06814318-f760-41e6-a124-cdada8c28f12)

**Reply by SeanCassiere** [reply]
@VladKopylov I agree disabling eslint is not the solution. However, to get around that eslint warning, you'll need to move the component out of the Route definition and reference it from there instead.

```tsx
export const Route = createFileRoute('/posts')({
	component: PostsComponent
})

function PostsComponent() {
	// ...
}
```
